servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/TimeoutFromRequest.java

```diff
@@ -37,10 +42,24 @@
      *
      * @param request the current request
      * @return The timeout or null for no timeout
+     * @deprecated Use {@link #apply(HttpRequestMetaData, TimeSource)}. This method will be removed in a future release.
      */
     @Override
+    @Deprecated
+    @Nullable
+    default Duration apply(HttpRequestMetaData request) {
+        return apply(request, (unit) -> unit.convert(nanoTime(), NANOSECONDS));
+    }
+
+    /**
+     * Determine timeout duration, if present, from a request and/or apply default timeout durations.
+     *
+     * @param request the current request
+     * @param timeSource {@link TimeSource} for calculating the timeout
+     * @return The timeout or null for no timeout
+     */
     @Nullable
-    Duration apply(HttpRequestMetaData request);
+    Duration apply(HttpRequestMetaData request, TimeSource timeSource);
```

Comment:
Note this is a breaking change and hard to avoid due to the nature of `@FunctionalInterface` from `java.util.function.Function`. This change is required in order to make the gRPC deadline use `TimeSource` instead of `System.nanoTime()`. We can discard the latest commit if we don't want to introduce this breaking change.

Reply:
This is similar to the case we had with a new method on LB interface. We can still do that without breaking change, it will just require one more release:
1. add new default method, keep the pre-existing as-is in 0.41;
2. swap default methods in 0.42;
3. kill old method in 0.43;

Reply:
I considered it. The drawbacks are the following:
* Users of 0.41 who currently provide a lambda, in order to get rid of the deprecated method, would need to swap the lambda with an actual implementation of the interface, and would still need to implement the non-default method,
* If a user ignores the deprecation, upon bumping to 0.42 the lambda won't work as the new method takes 2 arguments, and the 1-argument variant no longer is subject to the semantics of a functional interface. They will be forced to migrate.

As a result, that sounds like a breaking change in 0.42 is either the same amount of work for the user or less, depending on user's strategy. WDYT?

Reply:
Also, I'm not sure how frequently this interface is incorporated into codebases by end users.

The only surfacing area I see is:

`TimeoutHttpRequesterFilter` and `TimeoutHttpServiceFilter` constructors with `TimeoutFromRequest` as an argument -> that we could deprecate and provide overrides whose arguments are:

```
Function<HttpRequestMetaData, Duration> // deprecated, applies default strategy
Function<HttpRequestMetaData, Duration>, HttpExecutionStrategy strategy // deprecated
BiFunction<HttpRequestMetadata, TimeSource, Duration> // applies default strategy
BiFunction<HttpRequestMetadata, TimeSource, Duration>, HttpExecutionStrategy strategy
```

but this strategy is also problematic, as once we change the signatures in `TimeoutFromRequest` later we will introduce binary incompatibility even though source compatibility would be maintained. I'm just trying to come up with a smooth transition, but this also sounds as inconvenient.